File: generate_configs.py
import yaml
import logging
import os
from dotenv import load_dotenv
import shutil

# Charger les variables d'environnement
load_dotenv()

DOMAIN = os.getenv('DOMAIN')
ENV = os.getenv('ENV', 'production')
NETWORK_NAME = 'proxy-network'
ENTRYPOINT_YML_PATH = 'custom.yml'
OUTPUT_DIR = 'dist'  # Dossier de génération des fichiers

if DOMAIN is None:
    raise RuntimeError("The DOMAIN environment variable is not set.")

SERVICE_KEYS = {
    'required': {'image'},
    'optional': {'title', 'description', 'icon', 'environment', 'volumes', 'labels', 'subdomain', 'internal_port', 'ports', 'command', 'depends_on', 'storage', 'dev_path', 'group'}
}

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

class ConfigGenerator:
    def __init__(self, yml_path):
        self.yml_path = yml_path
        self.yml_data = self._read_yaml_file()
        if self.yml_data:
            self.docker_compose = {
                'services': {},
                'volumes': {},
                'networks': {
                    NETWORK_NAME: {
                        'external': True
                    }
                }
            }
            os.makedirs(OUTPUT_DIR, exist_ok=True)

            # Copier le fichier .env dans le dossier de sortie
            self.copy_env_file()

            docker_compose_yml = self.generate_docker_compose()
            self.write_file(os.path.join(OUTPUT_DIR, 'docker-compose.yml'), docker_compose_yml)

            # Générer Caddyfile
            caddyfile_content = self.generate_caddyfile()
            self.write_file(os.path.join(OUTPUT_DIR, 'Caddyfile'), caddyfile_content)

            # Générer services.yml
            services_yml = self.generate_services_yml()
            self.write_file(os.path.join(OUTPUT_DIR, 'homepage_services.yaml'), services_yml)

            # Générer proxy.docker-compose.yml
            proxy_docker_compose_yml = self.generate_proxy_docker_compose()
            self.write_file(os.path.join(OUTPUT_DIR, 'proxy.docker-compose.yml'), proxy_docker_compose_yml)

            logging.info("Files generated successfully!")

    def _read_yaml_file(self):
        if not os.path.exists(self.yml_path):
            logging.error(f"File not found: {self.yml_path}")
            return None
        try:
            with open(self.yml_path, 'r') as file:
                return yaml.safe_load(file)
        except yaml.YAMLError as exc:
            logging.error(f"Error parsing YAML file: {self.yml_path}\n{exc}")
            return None

    def write_file(self, filepath, content):
        os.makedirs(os.path.dirname(filepath), exist_ok=True)
        try:
            with open(filepath, 'w') as file:
                file.write(content)
                logging.info(f"File written successfully: {filepath}")
        except IOError as exc:
            logging.error(f"Error writing file: {filepath}\n{exc}")

    def copy_env_file(self):
        src = '.env'
        dst = os.path.join(OUTPUT_DIR, '.env')
        try:
            shutil.copy(src, dst)
            logging.info(f"Copied .env file to {dst}")
        except IOError as exc:
            logging.error(f"Error copying .env file: {src}\n{exc}")

    def _get_reverse_proxy_details(self, service):
        subdomain = service.get('subdomain', '')
        internal_port = service.get('internal_port', 80)
        return subdomain, internal_port

    def generate_docker_compose(self):
        for service_name, service in self.yml_data.get('services', {}).items():
            if not SERVICE_KEYS['required'].issubset(service):
                logging.warning(f"Service definition incomplete: {service_name}")
                continue

            compose_service_item = {
                'container_name': service_name,
                'image': service['image'],
                'restart': 'always',
                'networks': [NETWORK_NAME]
            }

            if 'environment' in service:
                compose_service_item['environment'] = [f"{var}=${{{var}}}" for var in service['environment']]
            if 'volumes' in service:
                compose_service_item['volumes'] = service['volumes']
            if 'labels' in service:
                compose_service_item['labels'] = service['labels']
            if 'ports' in service:
                compose_service_item['ports'] = service['ports']
            if 'command' in service:
                compose_service_item['command'] = service['command']
            if 'depends_on' in service:
                compose_service_item['depends_on'] = service['depends_on']

            # Create storage volume
            if 'storage' in service:
                volume_name = f"{service_name}_data"
                if service['storage'] == 'internal':
                    self.docker_compose['volumes'][volume_name] = {}
                else:
                    self.docker_compose['volumes'][volume_name] = {'external': True}
                compose_service_item['volumes'] = compose_service_item.get('volumes', []) + [f"{volume_name}:/data"]

            # Add dev_path if ENV is 'development'
            if ENV == 'development' and 'dev_path' in service:
                compose_service_item['volumes'] = compose_service_item.get('volumes', []) + [f"{service['dev_path']}:/app"]

            self.docker_compose['services'][service_name] = compose_service_item

            # Check for extra keys
            extra_keys = set(service.keys()) - SERVICE_KEYS['required'] - SERVICE_KEYS['optional']
            if extra_keys:
                logging.warning(f"Service '{service_name}' has extra keys not being used: {extra_keys}")

        return yaml.dump(self.docker_compose, sort_keys=False)

    def generate_proxy_docker_compose(self):
        proxy_docker_compose = {
            'services': {
                'caddy': {
                    'container_name': 'caddy',
                    'image': 'caddy:latest',
                    'volumes': [
                        '/var/run/docker.sock:/var/run/docker.sock',
                        './Caddyfile:/etc/caddy/Caddyfile'
                    ],
                    'ports': [
                        '80:80',
                        '443:443',
                        '443:443/udp',
                        '5000:5000'
                    ],
                    'restart': 'always',
                    'networks': [
                        NETWORK_NAME
                    ]
                }
            },
            'networks': {
                NETWORK_NAME: {
                    'external': True
                }
            }
        }
        return yaml.dump(proxy_docker_compose, sort_keys=False)

    def generate_caddyfile(self):
        lines = []
        base_config = self.yml_data.get('caddy_base_config', '')
        if base_config:
            lines.append(base_config)

        for service_name, service in self.yml_data.get('services', {}).items():
            subdomain, internal_port = self._get_reverse_proxy_details(service)
            if subdomain:
                lines.append(f"{subdomain}.{DOMAIN} {{")
            else:
                lines.append(f"{DOMAIN} {{")
            lines.append(f"    reverse_proxy {service_name}:{internal_port}")
            lines.append("}")

        return "\n".join(lines)

    def generate_services_yml(self):
        services_dict = {}
        groups = self.yml_data.get('groups', {})
        for service_name, service in self.yml_data.get('services', {}).items():
            group_id = service.get('group')
            group_name = groups.get(group_id, 'Ungrouped')
            subdomain, _ = self._get_reverse_proxy_details(service)
            href = f"http://{subdomain}.{DOMAIN}" if subdomain else f"http://{DOMAIN}"

            if group_name not in services_dict:
                services_dict[group_name] = []

            service_entry = {
                'href': href,
                'container': service_name,
                'showStats': True
            }

            if 'description' in service:
                service_entry['description'] = service['description']
            if 'icon' in service:
                service_entry['icon'] = service['icon']

            service_name_or_title = service.get('title', service_name)
            services_dict[group_name].append({service_name_or_title: service_entry})

        grouped_services = []
        for group, services in services_dict.items():
            grouped_services.append({group: services})
        return yaml.dump(grouped_services, sort_keys=False)

if __name__ == "__main__":
    generator = ConfigGenerator(ENTRYPOINT_YML_PATH)

File: generate_configs.js
import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';
import dotenv from 'dotenv';
import winston from 'winston';

// Charger les variables d'environnement
dotenv.config();

const DOMAIN = process.env.DOMAIN;
const ENV = process.env.ENV || 'production';
const NETWORK_NAME = 'proxy-network';
const ENTRYPOINT_YML_PATH = 'custom.yml';
const OUTPUT_DIR = 'dist';  // Dossier de génération des fichiers

if (!DOMAIN) {
    throw new Error("The DOMAIN environment variable is not set.");
}

const SERVICE_KEYS = {
    required: ['image'],
    optional: ['title', 'description', 'icon', 'environment', 'volumes', 'labels', 'subdomain', 'internal_port', 'ports', 'command', 'depends_on', 'storage', 'dev_path', 'group']
};

// Configurer le logging
const logger = winston.createLogger({
    level: 'info',
    format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.printf(({ timestamp, level, message }) => `${timestamp} - ${level} - ${message}`)
    ),
    transports: [
        new winston.transports.Console()
    ]
});

class ConfigGenerator {
    constructor(ymlPath) {
        this.ymlPath = ymlPath;
        this.ymlData = this._readYamlFile();
        if (this.ymlData) {
            this.dockerCompose = {
                services: {},
                volumes: {},
                networks: {
                    [NETWORK_NAME]: {
                        external: true
                    }
                }
            };
            fs.mkdirSync(OUTPUT_DIR, { recursive: true });

            // Copier le fichier .env dans le dossier de sortie
            this.copyEnvFile();

            const dockerComposeYml = this.generateDockerCompose();
            this.writeFile(path.join(OUTPUT_DIR, 'docker-compose.yml'), dockerComposeYml);

            // Générer Caddyfile
            const caddyfileContent = this.generateCaddyfile();
            this.writeFile(path.join(OUTPUT_DIR, 'Caddyfile'), caddyfileContent);

            // Générer services.yml
            const servicesYml = this.generateServicesYml();
            this.writeFile(path.join(OUTPUT_DIR, 'homepage_services.yaml'), servicesYml);

            // Générer proxy.docker-compose.yml
            const proxyDockerComposeYml = this.generateProxyDockerCompose();
            this.writeFile(path.join(OUTPUT_DIR, 'proxy.docker-compose.yml'), proxyDockerComposeYml);

            logger.info("Files generated successfully!");
        }
    }

    _readYamlFile() {
        if (!fs.existsSync(this.ymlPath)) {
            logger.error(`File not found: ${this.ymlPath}`);
            return null;
        }
        try {
            const fileContent = fs.readFileSync(this.ymlPath, 'utf8');
            return yaml.load(fileContent);
        } catch (error) {
            logger.error(`Error parsing YAML file: ${this.ymlPath}\n${error}`);
            return null;
        }
    }

    writeFile(filepath, content) {
        fs.mkdirSync(path.dirname(filepath), { recursive: true });
        try {
            fs.writeFileSync(filepath, content);
            logger.info(`File written successfully: ${filepath}`);
        } catch (error) {
            logger.error(`Error writing file: ${filepath}\n${error}`);
        }
    }

    copyEnvFile() {
        const src = '.env';
        const dst = path.join(OUTPUT_DIR, '.env');
        try {
            fs.copyFileSync(src, dst);
            logger.info(`Copied .env file to ${dst}`);
        } catch (error) {
            logger.error(`Error copying .env file: ${src}\n${error}`);
        }
    }

    _getReverseProxyDetails(service) {
        const subdomain = service.subdomain || '';
        const internalPort = service.internal_port || 80;
        return [subdomain, internalPort];
    }

    generateDockerCompose() {
        for (const [serviceName, service] of Object.entries(this.ymlData.services || {})) {
            if (!SERVICE_KEYS.required.every(key => key in service)) {
                logger.warn(`Service definition incomplete: ${serviceName}`);
                continue;
            }

            const composeServiceItem = {
                container_name: serviceName,
                image: service.image,
                restart: 'always',
                networks: [NETWORK_NAME]
            };

            if (service.environment) {
                composeServiceItem.environment = service.environment.map(varName => `${varName}=${process.env[varName]}`);
            }
            if (service.volumes) {
                composeServiceItem.volumes = service.volumes;
            }
            if (service.labels) {
                composeServiceItem.labels = service.labels;
            }
            if (service.ports) {
                composeServiceItem.ports = service.ports;
            }
            if (service.command) {
                composeServiceItem.command = service.command;
            }
            if (service.depends_on) {
                composeServiceItem.depends_on = service.depends_on;
            }

            // Create storage volume
            if (service.storage) {
                const volumeName = `${serviceName}_data`;
                if (service.storage === 'internal') {
                    this.dockerCompose.volumes[volumeName] = {};
                } else {
                    this.dockerCompose.volumes[volumeName] = { external: true };
                }
                composeServiceItem.volumes = [...(composeServiceItem.volumes || []), `${volumeName}:/data`];
            }

            // Add dev_path if ENV is 'development'
            if (ENV === 'development' && service.dev_path) {
                composeServiceItem.volumes = [...(composeServiceItem.volumes || []), `${service.dev_path}:/app`];
            }

            this.dockerCompose.services[serviceName] = composeServiceItem;

            // Check for extra keys
            const extraKeys = Object.keys(service).filter(key => !SERVICE_KEYS.required.includes(key) && !SERVICE_KEYS.optional.includes(key));
            if (extraKeys.length) {
                logger.warn(`Service '${serviceName}' has extra keys not being used: ${extraKeys}`);
            }
        }

        return yaml.dump(this.dockerCompose, { sortKeys: false });
    }

    generateProxyDockerCompose() {
        const proxyDockerCompose = {
            services: {
                caddy: {
                    container_name: 'caddy',
                    image: 'caddy:latest',
                    volumes: [
                        '/var/run/docker.sock:/var/run/docker.sock',
                        './Caddyfile:/etc/caddy/Caddyfile'
                    ],
                    ports: [
                        '80:80',
                        '443:443',
                        '443:443/udp',
                        '5000:5000'
                    ],
                    restart: 'always',
                    networks: [
                        NETWORK_NAME
                    ]
                }
            },
            networks: {
                [NETWORK_NAME]: {
                    external: true
                }
            }
        };
        return yaml.dump(proxyDockerCompose, { sortKeys: false });
    }

    generateCaddyfile() {
        const lines = [];
        const baseConfig = this.ymlData.caddy_base_config || '';
        if (baseConfig) {
            lines.push(baseConfig);
        }

        for (const [serviceName, service] of Object.entries(this.ymlData.services || {})) {
            const [subdomain, internalPort] = this._getReverseProxyDetails(service);
            if (subdomain) {
                lines.push(`${subdomain}.${DOMAIN} {`);
            } else {
                lines.push(`${DOMAIN} {`);
            }
            lines.push(`    reverse_proxy ${serviceName}:${internalPort}`);
            lines.push('}');
        }

        return lines.join('\n');
    }

    generateServicesYml() {
        const servicesDict = {};
        const groups = this.ymlData.groups || {};
        for (const [serviceName, service] of Object.entries(this.ymlData.services || {})) {
            const groupId = service.group;
            const groupName = groups[groupId] || 'Ungrouped';
            const [subdomain] = this._getReverseProxyDetails(service);
            const href = subdomain ? `http://${subdomain}.${DOMAIN}` : `http://${DOMAIN}`;

            if (!servicesDict[groupName]) {
                servicesDict[groupName] = [];
            }

            const serviceEntry = {
                href,
                container: serviceName,
                showStats: true
            };

            if (service.description) {
                serviceEntry.description = service.description;
            }
            if (service.icon) {
                serviceEntry.icon = service.icon;
            }

            const serviceNameOrTitle = service.title || serviceName;
            servicesDict[groupName].push({ [serviceNameOrTitle]: serviceEntry });
        }

        const groupedServices = [];
        for (const [group, services] of Object.entries(servicesDict)) {
            groupedServices.push({ [group]: services });
        }
        return yaml.dump(groupedServices, { sortKeys: false });
    }
}

const generator = new ConfigGenerator(ENTRYPOINT_YML_PATH);